### PR TITLE
feat: plumber pdf segmenter

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
This PR intends to use a more advanced PDF extractor library - pdf plumber to extract text. This is used for PDF search, and may also add more functions from this library, like position information, etc.

- [x] use [PDFplumber](https://pypi.org/project/pdfplumber/) to extract text as one chunk
- [x] use [pdftitle]( https://pypi.org/project/pdftitle/) to extract title as one chunk

### CODE REVIEW

File: segmenters/nlp/PDFPlumberSegmenter/__init__.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+        :param uri: File name of PDF
+        :type uri: str
+        :param buffer: PDF file in bytes
+        :type buffer: bytes
+        :param mime_type: the type of data
+        :returns: A list of documents with the extracted data
+        :rtype: List[Dict]
+        """
+        chunks = []
```
Comment (JoanFM) [inline segmenters/nlp/PDFPlumberSegmenter/__init__.py:35]: If we need this, we should add a FliterQL driver before this executor, but this should not be part here in my opinion

File: segmenters/nlp/PDFPlumberSegmenter/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+So if we have a PDF with 2 images and text, we will have 3 chunks; one per each image and one per the text.
+
+
+## 🏃️ Flow
+First you define your flow here:
+    
+    f = Flow().add(uses='PDFExtractorSegmenter', array_in_pb=True)
+
+If you have a PDF with images and text, the ```array_in_pb=True``` is necessary.
```
Comment (JoanFM) [inline segmenters/nlp/PDFPlumberSegmenter/README.md:15]: why is this necessary?

File: segmenters/nlp/PDFPlumberSegmenter/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+And then we search in: 
+
+    with f:
+        f.search(input_fn=search_generator(path=path, buffer=None), on_done=validate_mix_fn)
```
Comment (JoanFM) [inline segmenters/nlp/PDFPlumberSegmenter/README.md:20]: input_fn is deprecated

**Review by JoanFM — DISMISSED** [review]

File: segmenters/nlp/PDFPlumberSegmenter/__init__.py (hunk 1 above)
Comment (Yongxuanzhang) [reply segmenters/nlp/PDFPlumberSegmenter/__init__.py:35]: You mean `chunks = []`? Other segments also have this. Or you mean the mimetype?

File: segmenters/nlp/PDFPlumberSegmenter/README.md (hunk 2 above)
Comment (Yongxuanzhang) [reply segmenters/nlp/PDFPlumberSegmenter/README.md:15]: I haven't updated the readme. This is a copy-paste from another pdf segmenter. 😄

File: segmenters/nlp/PDFPlumberSegmenter/__init__.py (hunk 1 above)
Comment (Yongxuanzhang) [reply segmenters/nlp/PDFPlumberSegmenter/__init__.py:35]: Ah yes, we can add a driver. But this is easier to use. And I plan to add some warning information here to suggest users to use driver to filter only pdf data.

Comment (JoanFM) [reply segmenters/nlp/PDFPlumberSegmenter/__init__.py:35]: I am not convinced with this. This should not be this executors responsability I think

Comment (Yongxuanzhang) [reply segmenters/nlp/PDFPlumberSegmenter/__init__.py:35]: How can we solve this situation? I have 3 segments working sequentially, 1 text, 1 image, and 1 pdf. If I use FliterQL at first for the text segmenter, how can I keep the query data of images and pdf?

Comment (JoanFM) [reply segmenters/nlp/PDFPlumberSegmenter/__init__.py:35]: Why do u have these Sequentially?

Comment (Yongxuanzhang) [reply segmenters/nlp/PDFPlumberSegmenter/__init__.py:35]: I tried to use them in parallel but will get issues. So I use the compound to stack these 3 together.

Comment (JoanFM) [reply segmenters/nlp/PDFPlumberSegmenter/__init__.py:35]: ok keep it like this, but is weird

Comment (Yongxuanzhang) [reply segmenters/nlp/PDFPlumberSegmenter/__init__.py:35]: Yeah, true. I also don't like it. 😂 
I came up with 2 ways for this case:

- Use the 3 segmenters in parallel and each equipped with a filter driver. This is more aligned with our design. But I got very strange issues and cannot reproduce each time. I will try to reproduce it later and maybe open an issue.
- Use the 3 segmenters in sequential, and add this check inside the segmenters.

File: segmenters/nlp/PDFPlumberSegmenter/__init__.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+class PDFPlumberSegmenter(BaseSegmenter):
+    """
+    :class:`PDFPlumberSegmenter` Extracts data (text and images) from PDF files.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @single(slice_nargs=3)
+    def segment(self, uri: str, buffer: bytes, mime_type: str, *args, **kwargs) -> List[Dict]:
```
Comment (maximilianwerk) [inline segmenters/nlp/PDFPlumberSegmenter/__init__.py:21]: Too much is happening in this function. Please refactor the text extraction and the image extraction in separate functions.

File: segmenters/nlp/PDFPlumberSegmenter/tests/test_pdfextractorcrafter.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+path_img = os.path.join(cur_dir, 'cats_are_awesome_img.pdf')
+
+with open(path_text, 'rb') as pdf:
+    input_bytes_text = pdf.read()
+
+with open(path_img, 'rb') as pdf:
+    input_bytes_image = open(path_img, 'rb').read()
+
+with open(path_img_text, 'rb') as pdf:
+    input_bytes_images_text = pdf.read()
```
Comment (maximilianwerk) [inline segmenters/nlp/PDFPlumberSegmenter/tests/test_pdfextractorcrafter.py:25]: This code should go into a fixture and not on module-level.

**Review by maximilianwerk — DISMISSED** [review]
Content is good. Anyhow, I'd like some code restructuring for better maintainability.

File: segmenters/nlp/PDFPlumberSegmenter/README.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+PDFPlumberSegmenter is a segmenter used for extracting images and text as chunks from PDF data. It stores each images and text of each page as chunks separately.
+
+## Usage
+
+We use Pod images in several ways:
+
+1. Run with Docker: `docker run`
+   ```bash
+    docker run jinahub/pod.segmenter.pdfplumbersegmenter:0.0.1-1.0.1 --port-in 55555 --port-out 55556
```
Comment (cristianmtr) [inline segmenters/nlp/PDFPlumberSegmenter/README.md:11]: jina version is diff from dockerfile

Comment (Yongxuanzhang) [reply segmenters/nlp/PDFPlumberSegmenter/README.md]: fixed, thanks!

File: segmenters/nlp/PDFPlumberSegmenter/tests/test_pdfextractorcrafter.py (hunk 5 above)
Comment (Yongxuanzhang) [reply segmenters/nlp/PDFPlumberSegmenter/tests/test_pdfextractorcrafter.py]: Moved to fixture

File: segmenters/nlp/PDFPlumberSegmenter/__init__.py (hunk 4 above)
Comment (Yongxuanzhang) [reply segmenters/nlp/PDFPlumberSegmenter/__init__.py:57]: refactored, thx!

**Review by nan-wang — APPROVED** [review]
LGTM👍
